<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <arg name="world" default="empty"/> 
  <arg name="world_name" default="emty"/> 
  <arg name="paused" default="false"/>
  <arg name="use_sim_time" default="true"/>
  <arg name="gui" default="true"/>
  <arg name="headless" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="verbose" value="false" />
  <arg name="kinect_enabled" default="false"/>
  
<include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find cpmbot)/gazebo/worlds/$(arg world_name).world"/>
    <arg name="paused" value="$(arg paused)"/>
    <arg name="use_sim_time" value="$(arg use_sim_time)"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="headless" value="$(arg headless)"/>
    <arg name="debug" value="$(arg debug)"/>
    <arg name="verbose" value="$(arg verbose)"/>
</include>

<!-- load robot model -->
  <param name="robot_description" command="$(find xacro)/xacro.py '$(find cpmbot)/description/cpmbot.xacro'
    kinect_enabled:=$(arg kinect_enabled)"
     />

<node name="spawn_cpmbot_model" pkg="gazebo_ros" type="spawn_model"
    args=" -unpause -urdf -param robot_description -model cpmbot"/>

<!-- joy node -->   
<node respawn="true" pkg="joy" type="joy_node" name="joy_node" >
  <param name="dev" type="string" value="/dev/input/js0" />
  <param name="deadzone" value="0.12" />
</node>


<!-- Load joystick and operator configurations from YAML file to parameter server -->
<rosparam file="$(find cpmbot)/config/operator.yaml" command="load"/>
<node pkg="cpmbot" name="operator_node" type="operator_node"/>

<!-- Load basic movement configurations from YAML file to parameter server -->
<rosparam file="$(find cpmbot)/config/basic_movement.yaml" command="load"/>
<node pkg="cpmbot" type="basic_movement_node" name="basic_movement_node" />

<!-- remap mecanum commands node -->   
<param name="sub_front_left" type="string" value="controller/track_front_left/command" />
<param name="pub_front_left" type="string" value="controller/mecanum_front_left/command" />
<param name="sub_front_right" type="string" value="controller/track_front_right/command" />
<param name="pub_front_right" type="string" value="controller/mecanum_front_right/command" />
<param name="sub_back_left" type="string" value="controller/track_back_left/command" />
<param name="pub_back_left" type="string" value="controller/mecanum_back_left/command" />
<param name="sub_back_right" type="string" value="controller/track_back_right/command" />
<param name="pub_back_right" type="string" value="controller/mecanum_back_right/command" />
<!--node pkg="cpmbot" type="sub_pub_node" name="remap_mecanum_commands_node" >
</node-->


<!-- Load joint controller configurations from YAML file to parameter server -->
<rosparam file="$(find cpmbot)/config/gazebo_control.yaml" command="load"/>

<!-- load the controllers -->
<!-- joint_state_controller necesita ser instalado -->
  <node name="controller_manager_spawner" pkg="controller_manager" type="spawner" respawn="false"
    output="screen" ns="controller" args="joint_state_controller
  arm_front_left  
  arm_front_right
  arm_back_left
  arm_back_right
  track_front_left
  track_front_right
  track_back_left
  track_back_right"/>
<!--  mecanum_front_left
  mecanum_front_right
  mecanum_back_left
  mecanum_back_right
"/-->

 <!--node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" output="screen">
  </node-->
<!--param name="robot_description" command="$(find xacro)/xacro.py '$(find cpmbot)/description/cpmbot.xacro'"-->


  <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" respawn="false" output="screen">
    <!--param name="robot_description" command="$(find xacro)/xacro.py '$(find mybot_description)/urdf/mybot.xacro'"/-->
    <remap from="/joint_states" to="controller/joint_states" />
  </node>

    <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_node" clear_params="true">
    <rosparam command="load" file="$(find cpmbot)/config/localization.yaml" />

   <!-- Placeholder for output topic remapping
    <remap from="odometry/filtered" to=""/>
    <remap from="accel/filtered" to=""/>
    -->
  </node>


</launch>
